File: nunaliit2-js/src/main/webapp/nunaliit2/contributions.js
/*
Copyright (c) 2010, Geomatics and Cartographic Research Centre, Carleton 
University
All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are met:

 - Redistributions of source code must retain the above copyright notice, 
   this list of conditions and the following disclaimer.
 - Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
 - Neither the name of the Geomatics and Cartographic Research Centre, 
   Carleton University nor the names of its contributors may be used to 
   endorse or promote products derived from this software without specific 
   prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
POSSIBILITY OF SUCH DAMAGE.

$Id: contributions.js 8165 2012-05-31 13:14:37Z jpfiset $
*/

// @requires n2.utils.js

NUNALIIT_CONTRIBUTIONS = null;

function olkitContributions($, options_) { // defined here - but explicitly invoked by initContributionHandler() below

	// Localization
	var _loc = function(str){ return $n2.loc(str,'nunaliit2'); };

// === UTILITY ====================================	

	var _formIntroString =
		'Contribute information about this place to the atlas. ' +
		'Your contribution will appear along with other comments about ' +
		'this place shown in the side panel. You can also upload ' +
		'an audio recording, video or images about the place.';
	var _formUpdateNotice = 'If you proceed you will <i>UPDATE the following comment</i>.';
	var _formAddNotice = 'If you proceed you will <i>ADD a new comment</i>.';
	var defaults = {
		db: null
		,putUrl: 'upload/put'
		,deleteUrl: 'contributions/delete'
		,table: 'contributions'
		,anonymousAllowed: true
		,formData: {
			formIntroText: _formIntroString
			,updateNotice: _formUpdateNotice
			,addNotice: _formAddNotice
			,editFields: [  // code specific to these fields exists to create the form - these are the defaults
				{
					name: 'title'
					,label: 'Title'
					,type: 'text'
				}
				,{
					name: 'notes'
					,label: 'Comment'
					,type: 'textarea'
					,cols: '50'
					,rows: '10'
				}
				,{
					name: 'filename'
					,label: '' // this gets filled in
					,type: 'file'
				}
// sample date field - not used in default atlas
//				,{
//					name: 'sort_date'
//					,label: 'Earliest date'
//					,type: 'date'
//					,dpOptionsFn: function() { // return obj containing datepicker options - seems convoluted but this delays evaluation }
//				}
			]
		}
		,data: {} // used dynamically to carry restful parms - don't initialize programmatically
		,onSuccess: function(msg, textStatus){}
		,onError: function(xmlHttpRequest, textStatus, errorThrown){}
	};
	var contributionOptions = $.extend(true, {}, defaults, options_)
	
	var index = 0;
	var dialog = null;
	var form = null;

	function UserSubmit() {
		form.remove();
		form.css({display:"none",postion:"absolute",top:"0px",left:"0px"});
		//holder.append(form);
		$(document.body).append(form);
		dialog.dialog('close');

		var myForm = form;
		form = null;

		_SubmitForm(myForm);
	};

	function UserCancel() {
		dialog.dialog('close');
		if( form ) {
			form.remove();
			form = null;
		}
	};

	function _SubmitForm(myForm) {
		try {
			$(myForm).ajaxSubmit({
				beforeSubmit: BeforeFormSubmit
				,success: function(data,success){OnFormSuccess(data,success,myForm);}
				,error: function(xhr, status, e){OnFormError(xhr, status, e, myForm);}
				,iframe: true
			});
		} catch(error) {
			OnFormError(null, 'error', error, myForm);
		}
	};

	function _SubmitFormImmediate(options,form) {
		try {
			form.ajaxSubmit({
				success: function(data,success){ form.remove(); }
				,error: function(xhr, status, e){ form.remove(); }
				,iframe: true
			});
		} catch(error) {
			form.remove();
		};
	};
	
	function InstallKey(myForm,key) {
		$(myForm).attr("contributions-key",key);
		_SubmitForm(myForm);
	};

	function BeforeFormSubmit(parmArray,form,options) {
		var $form = $(form);
		
		var needKey = 1;
		
		if( ! $.progress ) {
			// There is no progress tracker, do not attempt
			// to get a key.
			needKey = 0;
		}
		
		var key = $form.attr("contributions-key");
		if( key ) {
			// There is a key, do not need to get one
			needKey = 0;
		}
		
		if( needKey ) {
			// Get key
			$.progress.getProgressKey(function(key){
				InstallKey(form,key);
			});
			
			// We need a key. Abort submitting.
			return false;
		}
		
		// At this point, we have a key or we do not need one
		if( key ) {
			$.progress.startProgressTrackingOn(key,'Submit contribution');
			
			$form.prepend('<input type="hidden" name="progressId" value="'+key+'"/>');
		};
		
		return true;
	};
	
	function OnFormSuccess(data, status, myForm) {
		//if( window.console && window.console.log ) window.console.log('OnCompletion',status,data);
		
		myForm.remove();
	};

	function OnFormError(xhr, status, e, myForm) {
		//if( window.console && window.console.log ) window.console.log('OnError',status,xhr,e);
		
		myForm.remove();
	};

	function clearFileFields(oObj) {
		oObj.filename = null;
		oObj.original_filename = null;
		oObj.mimetype = null;
		oObj.file_size = null;
	}
	
	function createForm(options) {
		var fileDivName = '____contrib_fileDiv';
		var fileDeleteDivName = '____contrib_fileDeleteDiv';
		var deleteCBName = '___contrib_deleteCB';
		
		/*
		 * If want multiple files attached, then will need to sort out field handling in the server as well
		 * for all of the fields derived from the upload: size, mimetype, etc.
		 *
		 * The assumptions about labelling made here are just part of the story.
		 */
		function insertFileReplaceDialog() {
			if (!options.data.deleteFile) {
				$('#'+fileDeleteDivName).html(
					'<span class="jqmContributions_input_heading">Replace attached file:</span>' +
						'<input class="jqmContributions_input" type="file" name="upload"/>');
			} else {
				$('#'+fileDeleteDivName).empty();
			}
		}
		
		function insertFileDiv() {
			var offerFileDelete = (typeof options.data.deleteFile != 'undefined');
			if (!offerFileDelete) { // no existing file - delete option not needed
				var htmlString = 
					'<span class="jqmContributions_input_heading">Attach file:</span>' +
						'<input class="jqmContributions_input" type="file" name="upload"/>';
			} else {
				var htmlString = 
					'<span class="jqmContributions_input_heading">Delete attached file:</span>' +
						'<input class="jqmContributions_input" type="checkbox" id="' + deleteCBName + '"/>' +
					'<div id="' + fileDeleteDivName + '"></div>';
			}
			$('#'+fileDivName).html(htmlString);
			
			if (offerFileDelete) {
				insertFileReplaceDialog();
				
				$('#'+deleteCBName).click(function() {
					if ($('#'+deleteCBName).attr('checked')) {
						options.data.deleteFile = true;
						clearFileFields(options.data);
					} else {
						options.data.deleteFile = false;
					}
					insertFileReplaceDialog();
				});
			}
		}
		
		function isEditField(name) {
			for (var i=0; i<options.formData.editFields.length; i++) {
				if (name == options.formData.editFields[i].name) {
					return(true);
				}
			}
			return(false);
		}
		
		function isFileDerivedField(name) {
			return(name == 'filename' ||
				name == 'mimetype' ||
				name == 'original_filename' ||
				name == 'file_size');
		}
		
		if( !form ) {
			var myIndex = index;
			++index;
			
			var hiddenData = '';
			for (var name in options.data) {
				if (!options.data.isUpdate || // these fields should not actually be there if not an update
						(!isEditField(name) &&
						 !isFileDerivedField(name) &&
						 name != 'deleteFile' &&
						 name != 'fileuse')) {
					var value = options.data[name];
					hiddenData += '<input type="hidden" name="'+name+'" value="'+value+'"/>';
				}
			};
			
			$(dialog).empty();
			var formId = 'contributionForm_'+myIndex;
			form = $('<form id="'+formId+'" action="'+options.putUrl+'" method="post" enctype="multipart/form-data">' + hiddenData + '</form>');
			$(dialog).append(form);
			form.append(
				'<p class="jqmContributions_instructions">' + options.formData.formIntroText + '</p>' +
				'<p class="jqmContributions_instructions">' +
					(options.data.isUpdate ? options.formData.updateNotice : options.formData.addNotice) +
				'</p>');
				
			// The name attributes in the form SHOULD MATCH the db schema column names
			// otherwise the server side has to contain hacks to translate it.  Use
			// whatever terminology you want in the accompanying label spans.  See 
			// comment/notes for example: the default db column name is notes but 
			// comment is used on the web form.
			for (var i=0; i < options.formData.editFields.length; i++) {
			
				/*
				 * create HTML
				 */
				var inputs = '';
				var iElem = $('<p></p>');
				form.append(iElem);
				
				if (options.formData.editFields[i].type != 'file') { // special handling for file labels
					iElem.append('<span class="jqmContributions_input_heading">' + options.formData.editFields[i].label + ':</span>');
				}
				
				if (options.formData.editFields[i].type == 'text') { // TEXT
				
					inputs =
						'<input class="jqmContributions_input" type="text" name="' +
						options.formData.editFields[i].name + '"/>';
						
				} else if (options.formData.editFields[i].type == 'textarea') { // TEXTAREA
				
					var _r = options.formData.editFields[i].rows;
					var _c = options.formData.editFields[i].cols;
					inputs =
						'<textarea class="jqmContributions_input" ' +
							'cols="' + (typeof _c != 'undefined' && _c != null && _c != '' ? _c : '50') + '" ' +
							'rows="' + (typeof _r != 'undefined' && _r != null && _r != '' ? _r : '10') + '" name="' +
							options.formData.editFields[i].name + '"></textarea>';
							
				} else if (options.formData.editFields[i].type == 'date') { // DATE - jQuery datepicker

						inputs = '<input id="__contrib_date_' + options.formData.editFields[i].name +
							'" class="jqmContributions_input" name="' + options.formData.editFields[i].name + '" type="text">';
				
				} else if (options.formData.editFields[i].type == 'file') { // FILE - attach/replace file dialog
				
						inputs = '<div id="' + fileDivName + '"></div>';
					
				} else { // NO MATCH - complain
					alert('contributions::createForm: unknown field type: '+options.formData.editFields[i].type);
				}
				iElem.append(inputs);
				
				/*
				 * init element if needed
				 */
				if ('file' == options.formData.editFields[i].type) { // FILE
				
					insertFileDiv();
			
				} else if ('date' == options.formData.editFields[i].type) { // DATE

					var dpBaseOpts = {
						dateFormat: 'yy-mm-dd'
						,gotoCurrent: true
						,changeYear: true
					};
					var dpOpt = $.extend({}, dpBaseOpts, options.formData.editFields[i].dpOptionsFn());
					$('#__contrib_date_' + options.formData.editFields[i].name).datepicker(dpOpt);
					
				}
				
			}
			form.append(
				'<p>' +
					'<input class="jqmContributions_input" id="contributionSubmit_'+myIndex+'" ' +
						'type="button" value="Submit"/>' +
					'<input class="jqmContributions_input" id="contributionCancel_'+myIndex+'" ' +
						'type="button" value="Cancel"/><br/>' +
				'</p>');
			
			if (options.data.isUpdate) { // fill the editable fields
				for (var i=0; i<options.formData.editFields.length; i++) {
					var efName = options.formData.editFields[i].name;
					if ('date' == options.formData.editFields[i].type) {
						if (isDefined(options.data[efName]) && isDefined(options.data[efName].formatted)) {
							$('#'+formId+' [name="'+efName+'"]').val(options.data[efName].formatted);
						}
					} else {
						$('#'+formId+' [name="'+efName+'"]').val(options.data[efName]);
					}
				};
			};

			$('#contributionSubmit_'+myIndex).click(function(evt){
				if (typeof options.data.deleteFile != 'undefined' && options.data.deleteFile) {
					$('#'+formId).append('<input type="hidden" name="deleteFile" value="'+options.data.deleteFile+'"/>');
					$('#'+formId).append('<input type="hidden" name="fileuse" value=""/>'); // also clear the file use flag if deleting a file
				} else {
					$('#'+formId).append('<input type="hidden" name="fileuse" value="'+
						(typeof options.data.fileuse == 'undefined' ? '' : options.data.fileuse) +'"/>');
				}
				UserSubmit();
			});
			$('#contributionCancel_'+myIndex).click(function(evt){
				UserCancel();
			});
		}
	};
	
	function createHiddenForm(options) {
		var form = $('<form action="'+options.putUrl+'" method="post" enctype="multipart/form-data"></form>');
		
		for(var name in options.data) {
			var value = options.data[name];

			var data = $('<input type="hidden" name="'+name+'"/>');
			form.append(data);
			data.attr('value',value);
		};
		
		return form;
	};
	
	function createDialog(options) {
		if( !dialog ) {
			dialog = $('<div></div>');
			$(dialog).dialog({
				autoOpen: false
				,modal: true
				,width: 'auto'
				,title: _loc( 'Contribution' )
				,close: function(event, ui){
					var diag = $(event.target);
					//diag.dialog('destroy');
					//diag.remove();
				}
			});
		}
		
		createForm(options);
	};

	return {
		acceptsContribution: function(options_) {
			var options = $.extend(true, {}, contributionOptions, options_);
			
			options.data.isUpdate = false;
			createDialog(options);
			$(dialog).dialog('open');
		}
		,acceptsContributionUpdate: function(options_) {
			var options = $.extend(true, {}, contributionOptions, options_);
			
			function setFileFields(offerDelete) {
				clearFileFields(options.data);
				if (offerDelete) {
					// by default, don't delete existing file but this will
					// cause the option to be offered.
					options.data.deleteFile = false;
				}
			}
			
			options.data.isUpdate = true;
			if (typeof options.data.filename != 'undefined' &&
				options.data.filename != null) {
				// for update - clear existing file fields.  Offer options to 
				// allow addition (if no file exists now) or delete/replace
				// if a file does exist.
				if (options.data.filename != '') {
					setFileFields(true);
				} else {
					setFileFields(false);
				}
			} else {
				setFileFields(false);
			}
			
			createDialog(options);
			$(dialog).dialog('open');
		}
		,deleteContribution: function(options_) {
			var options = $.extend(true, {}, contributionOptions, options_);
			
			$.ajax({
				type: 'GET'
				,url: options.deleteUrl
				,data: options.data
				,dataType: 'json'
				,async: true
				,success: options.onSuccess
				,error: options.onError
			});
		}
		,addContribution: function(options_) {
			var options = $.extend(true, {}, contributionOptions, options_);
			
			options.data.isUpdate = false;
			var form = createHiddenForm(options);
			$(document.body).append(form);
			_SubmitFormImmediate(options,form);
		}
		,allowAnonymousContribution: function() {
			return(contributionOptions.anonymousAllowed);
		}
	};
};

function initContributionHandler($, options_) {
	NUNALIIT_CONTRIBUTIONS = olkitContributions($, options_);
}

File: nunaliit2-contributions/src/main/java/ca/carleton/gcrc/contributions/Contributions.java
/*
Copyright (c) 2010, Geomatics and Cartographic Research Centre, Carleton 
University
All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are met:

 - Redistributions of source code must retain the above copyright notice, 
   this list of conditions and the following disclaimer.
 - Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.
 - Neither the name of the Geomatics and Cartographic Research Centre, 
   Carleton University nor the names of its contributors may be used to 
   endorse or promote products derived from this software without specific 
   prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
POSSIBILITY OF SUCH DAMAGE.

$Id$
*/
package ca.carleton.gcrc.contributions;

import ca.carleton.gcrc.auth.common.UserRepository;
import ca.carleton.gcrc.contributionsImpl.ContributionServerSupportedFieldsImpl;
import ca.carleton.gcrc.contributionsImpl.ContributionClientSideFieldsImpl;
import ca.carleton.gcrc.dbSec.impl.ColumnDataUtils;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.Statement;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Vector;

import javax.servlet.ServletException;

import org.apache.log4j.Logger;

import org.json.JSONArray;
import org.json.JSONObject;

public class Contributions {
	final protected Logger logger = Logger.getLogger(this.getClass());

	static final public String PROPERTIES_KEY_TABLE_NAME = "contributions.tableName";
	static final public String PROPERTIES_KEY_ID_COLUMN_NAME = "contributions.idColumnName";
	static final public String PROPERTIES_KEY_ID_SEQUENCE_NAME = "contributions.idSequenceName";
	static final public String PROPERTIES_KEY_SERVER_SUPPORTED_COLUMNS = "contributions.serverSupportedColumns";
	static final public String PROPERTIES_KEY_CLIENT_SIDE_COLUMNS = "contributions.clientSideColumns";
	
	static final public String DEFAULT_TABLE_NAME = "contributions";
	static final public String DEFAULT_ID_COLUMN_NAME = "id";
	static final public String DEFAULT_ID_SEQUENCE_NAME = "contributions_id_seq";
	static final public String DEFAULT_SERVER_SUPPORTED_COLUMNS =
		"id,place_id,contributor_id,create_ts,create_ms,last_edit_timestamp,last_edit_id,filename,original_filename,mimetype,file_size";
	static final public String DEFAULT_CLIENT_SIDE_COLUMNS =
		"title,notes,fileuse,likes,dislikes,related_to";
	
	private ContributionComet contributionComet = new ContributionCometNull();
	private Connection connection;
	private String tableName;
	private String idColumnName;
	private String idSequenceName;
	private String deleteSqlStatement;
	private ContributionsFieldSubset serverSupported;
	private ContributionsFieldSubset clientSide;
	
	public Contributions(Properties props, Connection connection) throws Exception {
		this.connection = connection;
		serverSupported = new ContributionServerSupportedFieldsImpl();
		clientSide = new ContributionClientSideFieldsImpl();

		readProperties(props);
	}

	private void readProperties(Properties props) throws Exception {
		tableName = props.getProperty(PROPERTIES_KEY_TABLE_NAME, DEFAULT_TABLE_NAME);
		idColumnName = props.getProperty(PROPERTIES_KEY_ID_COLUMN_NAME, DEFAULT_ID_COLUMN_NAME);
		idSequenceName = props.getProperty(PROPERTIES_KEY_ID_SEQUENCE_NAME, DEFAULT_ID_SEQUENCE_NAME);
		
		String serverFieldList = props.getProperty(PROPERTIES_KEY_SERVER_SUPPORTED_COLUMNS, DEFAULT_SERVER_SUPPORTED_COLUMNS);
		String clientFieldList = props.getProperty(PROPERTIES_KEY_CLIENT_SIDE_COLUMNS, DEFAULT_CLIENT_SIDE_COLUMNS);
		createServerSupportedAndClientSideLists(serverFieldList, clientFieldList);
		allFieldsInDb(serverFieldList, true); // verify all fields in prop lists accounted for in db - throw exception otherwise
		allFieldsInDb(clientFieldList, false);
		
		if (serverSupported.includes(idColumnName)) { // mark that column as auto incremented
			serverSupported.setAutoIncrementSequence(idColumnName, idSequenceName);
		} else {
			throw new Exception(idColumnName + " must be included in the serverSupported field list.");
		}
				
		deleteSqlStatement = "DELETE FROM \""+tableName+"\" WHERE \""+idColumnName+"\" = ?;";
	}
	
	private void createServerSupportedAndClientSideLists(String serverFieldList, String clientFieldList) throws Exception {
		String[] serverFields = serverFieldList.split(",");
		String[] clientFields = clientFieldList.split(",");
		
		String sqlQuery = null;
		{ // get full list of field names according to database schema
			StringWriter sw = new StringWriter();
			PrintWriter pw = new PrintWriter(sw);
			pw.print("SELECT * FROM "+tableName+" LIMIT 1;");
			pw.flush();
			
			sqlQuery = sw.toString();
		}
		
		Statement stmt = connection.createStatement();
		if (stmt.execute(sqlQuery)) {
			ResultSet rs = stmt.getResultSet();
			ResultSetMetaData rsmd = rs.getMetaData();
			
			int count = rsmd.getColumnCount();
			for (int loop=0; loop<count; ++loop) {
				String columnName = rsmd.getColumnName(loop+1);
				boolean isServerSupported = isColumnNameInList(columnName, serverFields);
				boolean isClientSide = isColumnNameInList(columnName, clientFields);
				
				if (isServerSupported && isClientSide) { // huh? in both lists
					throw new Exception(columnName + " is included in both the serverSupported and clientSide contribution field lists.");
				} else if (isServerSupported) {
					serverSupported.addColumn(columnName, rsmd.getColumnType(loop+1), rsmd.getColumnTypeName(loop+1));
				} else if (isClientSide) {
					clientSide.addColumn(columnName, rsmd.getColumnType(loop+1), rsmd.getColumnTypeName(loop+1));
				} else { // huh? in neither list
					throw new Exception(columnName + " is NOT included in either the serverSupported or clientSide contribution field lists.");
				}
			}	
		}
	}
	
	private boolean isColumnNameInList(String columnName, String[] list) {
		boolean ret = false;
		for (String column : list) {
			if (column.equals(columnName)) {
				return(true);
			}
		}
		return(ret);
	}
	
	private void allFieldsInDb(String list, boolean checkServerSupported) throws Exception {
		String[] fieldNames = list.split(",");
		for (String field : fieldNames) {
			boolean exists = false;
			
			if (checkServerSupported && serverSupported.includes(field)) {
				exists = true;
			} else if (!checkServerSupported && clientSide.includes(field)) {
				exists = true;
			}
			if (!exists) {
				throw new Exception(field + " does not exist in the " + tableName + " schema.");
			}
		}
	}
	
	public void deleteContribution(String contributionId, String placeId) throws Exception {
		// Contribution id is an integer
		int id = Integer.parseInt(contributionId);
		
		PreparedStatement stmt = connection.prepareStatement(deleteSqlStatement);
		stmt.setInt(1, id);
//logger.info(stmt.toString());
		stmt.execute();
		
		contributionComet.reportDeletedContribution(placeId, contributionId);
	}

	public JSONObject fromName(String intPlaceId, UserRepository userRepository) throws Exception {
		String queryString = "SELECT " + serverSupported.getFieldNamesList();
		String clientFieldList = clientSide.getFieldNamesList();
		
		/*
		 * Assume there must be serverSupported fields but allow for no clientSide fields...
		 */
		if (clientFieldList != "") {
			queryString += "," + clientFieldList;
		}
		queryString += " FROM " + tableName + " WHERE place_id=?;";
		
		PreparedStatement stmt = connection.prepareStatement(queryString);
		serverSupported.writeToPreparedStatement(stmt, "place_id", 1, intPlaceId);
//logger.info(stmt.toString());
		JSONArray contributions = executeContribSelectToJson(stmt, userRepository);
		JSONObject result = new JSONObject();
		result.put("contributions", contributions);
		return(result);
	}
	
	private JSONArray executeContribSelectToJson(
			PreparedStatement pstmt,
			UserRepository userRepository) throws Exception {
		Map<Integer,JSONObject> userInfoMap = new HashMap<Integer,JSONObject>();

		boolean resultAvailable = pstmt.execute();
		if (!resultAvailable) { // There's a ResultSet to be had
			// indicates an update count or no results - this must be no results
			throw new Exception("Query returned no results");
		}
		ResultSet rs = pstmt.getResultSet();
		ResultSetMetaData rsmd = rs.getMetaData();

		JSONArray array = new JSONArray();
		try {
			while (rs.next()) {
				
				JSONObject obj = new JSONObject();
				for (int loop=0; loop<rsmd.getColumnCount(); ++loop) {
					int index = loop+1;

					String columnName = rsmd.getColumnName(index);

					if ("contributor_id".equalsIgnoreCase(columnName) ||
							"last_edit_id".equalsIgnoreCase(columnName)) {
						/*
						 * instead of inserting the id, swap in an object containing the 
						 * contributor information referenced by that id.
						 */

						JSONObject userInfo = null;
						int contributor_id = rs.getInt(index);
						Integer contId = new Integer(contributor_id);

						if (userInfoMap.containsKey(contId)) {
							userInfo = userInfoMap.get(contId);
						} else {
							try {
								userInfo = userRepository.userInfoFromId(contributor_id);
							} catch(Exception e) {
								// just ignore. User might be deleted from database
							}
							userInfoMap.put(contId, userInfo); // Cache result
						}

						if (null != userInfo) { // swap in contributor info - client side needs to be written to look for these...
							if ("contributor_id".equalsIgnoreCase(columnName)) {
								obj.put("contributor", userInfo);
							} else {
								obj.put("lastContributor", userInfo);								
							}
						}

					} else { // insert column after figuring out type info
						ColumnDataUtils.addColumnToJson(obj, rs, index, columnName, rsmd.getColumnType(index), rsmd.getColumnTypeName(index));
					}
				}
				array.put(obj);
				
			}
		} catch (Exception je) {
			throw new ServletException("Error while parsing results",je);
		}

		return(array);
	}

	public void insert(Map<String, List<String>> fieldPairs) throws Exception {
		String serverFields = serverSupported.getFieldNamesList();
		String clientFields = clientSide.getFieldNamesList();
		
		if (serverSupported.includes(idColumnName)) { // for insert - need to autoincrement id
			int aiValue = ColumnDataUtils.obtainNextIncrementInteger(
					connection,
					serverSupported.getColumnData(idColumnName));
			List<String> paramList = fieldPairs.get(idColumnName);
			if (null == paramList) {
				paramList = new Vector<String>();
				fieldPairs.put(idColumnName, paramList);
			}
			paramList.add("" + aiValue);
		} else {
			throw new Exception(idColumnName + " not included in serverSupported field list.");
		}
		
		String queryString = "INSERT INTO " + tableName + " (" + serverFields;
		if (clientFields != "") {
			queryString += "," + clientFields;
		}
		queryString += ") VALUES (";
		
		String[] eTokens = serverFields.split(",");
		String[] oTokens = clientFields.split(",");
		{
			boolean first = true;
			for (int i = 0; i < eTokens.length; i++) {
				if (first) {
					first = false;
				} else {
					queryString += ",";
				}
				queryString += serverSupported.getInsertWildcard(eTokens[i]);
			}
		}
		{
			for (int i = 0; i < oTokens.length; i++) {
				queryString += ",";
				queryString += clientSide.getInsertWildcard(oTokens[i]);
			}
		}
		queryString += ");";

		PreparedStatement stmt = connection.prepareStatement(queryString);
		int index = 1;  // reprocess all fields in the same order...
		{
			for (int i = 0; i < eTokens.length; i++) {
				serverSupported.addParameterToPreparedStatement(stmt, eTokens[i], index, fieldPairs);
				index++;
			}
		}
		{
			for (int i = 0; i < oTokens.length; i++) {
				clientSide.addParameterToPreparedStatement(stmt, oTokens[i], index, fieldPairs);
				index++;
			}
		}
//logger.info(stmt.toString());
		stmt.execute();
	}
		
	public void update(Map<String, List<String>> fieldPairs) throws Exception {
		String[] fullList_serverFields = serverSupported.getFieldNamesList().split(",");
		String[] fullList_clientFields = clientSide.getFieldNamesList().split(",");
		
		String serverFields = "";
		{
			String[] dontChange_serverSupported = { "id", "place_id", "contributor_id", "create_ts", "create_ms" };
			boolean first = true;
			for (int i=0; i<fullList_serverFields.length; i++) {
				if (!isColumnNameInList(fullList_serverFields[i], dontChange_serverSupported)) {
					if (fieldPairs.containsKey(fullList_serverFields[i])) {
						if (first) {
							first = false;
						} else {
							serverFields += ",";
						}
						serverFields += fullList_serverFields[i];
					}
				}
			}
		}
		
		String clientFields = "";
		{
			boolean first = true;
			for (int i=0; i<fullList_clientFields.length; i++) {
				if (fieldPairs.containsKey(fullList_clientFields[i])) {
					if (first) {
						first = false;
					} else {
						clientFields += ",";
					}
					clientFields += fullList_clientFields[i];
				}
			}
		}
		
		String queryString = "UPDATE " + tableName + " SET (" + serverFields;
		if (clientFields != "") {
			queryString += "," + clientFields;
		}
		queryString += ") = (";
		
		String[] eTokens = serverFields.split(",");
		String[] oTokens = clientFields.split(",");
		{
			boolean first = true;
			for (int i = 0; i < eTokens.length; i++) {
				if (first) {
					first = false;
				} else {
					queryString += ",";
				}
				queryString += serverSupported.getUpdateWildcard(eTokens[i]);
			}
		}
		{
			for (int i = 0; i < oTokens.length; i++) {
				queryString += ",";
				queryString += clientSide.getUpdateWildcard(oTokens[i]);
			}
		}
		queryString += ") WHERE id = " + serverSupported.getUpdateWildcard("id");
		queryString += " AND place_id = " + serverSupported.getUpdateWildcard("place_id") + ";";

		PreparedStatement stmt = connection.prepareStatement(queryString);
		int index = 1;  // reprocess all fields in the same order...
		{
			for (int i = 0; i < eTokens.length; i++) {
				serverSupported.addParameterToPreparedStatement(stmt, eTokens[i], index, fieldPairs);
				index++;
			}
		}
		{
			for (int i = 0; i < oTokens.length; i++) {
				clientSide.addParameterToPreparedStatement(stmt, oTokens[i], index, fieldPairs);
				index++;
			}
		}
		serverSupported.addParameterToPreparedStatement(stmt, "id", index, fieldPairs);
		index++;
		serverSupported.addParameterToPreparedStatement(stmt, "place_id", index, fieldPairs);
		
//logger.info(stmt.toString());
		stmt.execute();
	}
		
	public String getTableName() {
		return tableName;
	}

	public String getIdColumnName() {
		return idColumnName;
	}

	public ContributionComet getContributionComet() {
		return contributionComet;
	}

	public void setContributionComet(ContributionComet contributionComet) {
		this.contributionComet = contributionComet;
	}

}
